feat: add accessor arrays support to `blas/ext/base/gcusumkbn2`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   add accessor arrays support to `blas/ext/base/gcusumkbn2`

## Related Issues

> Does this pull request have any related issues?

No. 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add accessor arrays support

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/lib/ndarray.js:
```suggestion
	if ( ox.accessorProtocol || oy.accessorProtocol ) {
```

headlessNode committed:
refactor: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/lib/accessors.js:
```suggestion
* gcusumkbn2( 4, 0.0, arraylike2object( x ), 2, 1, arraylike2object( y ), 1, 0 );
```

headlessNode committed:
docs: apply review suggestion

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/docs/types/index.d.ts:
```suggestion
* Output array.
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/docs/types/index.d.ts:
```suggestion
	<T extends OutputArray>( N: number, sum: number, x: InputArray, strideX: number, y: T, strideY: number ): T;
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/docs/types/index.d.ts:
```suggestion
	ndarray<T extends OutputArray>( N: number, sum: number, x: InputArray, strideX: number, offsetX: number, y: T, strideY: number, offsetY: number ): T;
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/docs/types/test.ts:
```suggestion
	gcusumkbn2( x.length, 0.0, x, 1, y, 1 ); // $ExpectType Float64Array
	gcusumkbn2( x.length, 0.0, new AccessorArray( x ), 1, new AccessorArray( y ), 1 ); // $ExpectType AccessorArray
```

headlessNode commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/docs/types/test.ts:
```suggestion
	gcusumkbn2.ndarray( x.length, 0.0, x, 1, 0, y, 1, 0 ); // $ExpectType Float64Array
	gcusumkbn2.ndarray( x.length, 0.0, new AccessorArray( x ), 1, 0, new AccessorArray( y ), 1, 0 ); // $ExpectType AccessorArray
```

headlessNode committed:
refactor: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/lib/ndarray.js:
```suggestion
		accessors( N, sum, ox, strideX, offsetX, oy, strideY, offsetY );
		return y;
```

You need to return the original output array, not the wrapped object.

kgryte committed:
fix: update return value

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/test/test.main.js:
@headlessNode I suggest adding a test case here when providing accessor arrays. This would have caught the bug mentioned above.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/test/test.ndarray.js:
As an aside, the manual loop is fine with me, but, for reference, we also recently added [`isSameArray`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/assert/is-same-array), which abstracts array such manual iteration.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/gcusumkbn2/test/test.ndarray.js:
@headlessNode Same comment as above wrt adding an accessor array test.

kgryte reviewed (changes requested):
Other than a couple of suggested test cases, this is looking good to me.

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte reviewed (approved):
LGTM
